From 033be94559a71372a7668616fb561b36cd1a814c Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Mon, 22 Dec 2025 09:12:05 +0100
Subject: [PATCH] omp: avoid taking address of reference in map

For reference parameters, lower_omp_target was using build_fold_addr_expr on
the decl when computing hostaddrs for a mapping clause.  For a reference this
produces the address of the reference slot, which may live in the current stack
frame, instead of the referenced object address.

Use the reference value (converted to void *) instead.

	PR middle-end/103276

gcc/ChangeLog:

	* omp-low.cc (lower_omp_target): Avoid build_fold_addr_expr for
	reference parameters; use the reference value.

gcc/testsuite/ChangeLog:

	* gfortran.dg/goacc/pr103276-enter-data-refmap.f90: New test.

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/omp-low.cc                                | 10 ++++--
 .../goacc/pr103276-enter-data-refmap.f90      | 36 +++++++++++++++++++
 2 files changed, 44 insertions(+), 2 deletions(-)
 create mode 100644 gcc/testsuite/gfortran.dg/goacc/pr103276-enter-data-refmap.f90

diff --git a/gcc/omp-low.cc b/gcc/omp-low.cc
index 4540a25d1ad..07fb27ad2b6 100644
--- a/gcc/omp-low.cc
+++ b/gcc/omp-low.cc
@@ -6937,7 +6937,8 @@ lower_rec_input_clauses (tree clauses, gimple_seq *ilist, gimple_seq *dlist,
 
 	  gimple_seq_add_seq (dlist, llist[2]);
 
-	  g = gimple_build_assign (simt_lane, LSHIFT_EXPR, simt_lane, integer_one_node);
+		  g = gimple_build_assign (simt_lane, LSHIFT_EXPR, simt_lane,
+					   integer_one_node);
 	  gimple_seq_add_stmt (dlist, g);
 
 	  gimple_seq_add_stmt (dlist, gimple_build_label (header));
@@ -13465,7 +13466,12 @@ lower_omp_target (gimple_stmt_iterator *gsi_p, omp_context *ctx)
 			      && OMP_CLAUSE_CODE (c) != OMP_CLAUSE_TO)
 			     || (!omp_is_allocatable_or_ptr (var)
 				 && !omp_check_optional_argument (var, false)))
-		      var = build_fold_addr_expr (var);
+		      {
+			if (TREE_CODE (TREE_TYPE (ovar)) == REFERENCE_TYPE)
+			  var = fold_convert_loc (loc, ptr_type_node, var);
+			else
+			  var = build_fold_addr_expr (var);
+		      }
 		    gimplify_assign (x, var, &ilist);
 		  }
 	      }
diff --git a/gcc/testsuite/gfortran.dg/goacc/pr103276-enter-data-refmap.f90 b/gcc/testsuite/gfortran.dg/goacc/pr103276-enter-data-refmap.f90
new file mode 100644
index 00000000000..0e72965660e
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/goacc/pr103276-enter-data-refmap.f90
@@ -0,0 +1,36 @@
+! { dg-do compile }
+! { dg-options "-O2 -fopenacc -fdump-tree-omplower" }
+!
+! PR 103276
+! OpenACC ENTER DATA on a derived-type dummy argument is lowered using an
+! internal reference parameter.  The mapping must not use the address of the
+! reference slot, which can point into the current stack frame.
+
+module m
+  use, intrinsic :: iso_fortran_env, only: dp => real64
+  implicit none
+
+  type :: des_t
+    real(dp) :: a
+  end type des_t
+
+contains
+
+  subroutine copyin_simple (var)
+    type(des_t), intent(inout) :: var
+!$acc enter data copyin(var)
+  end subroutine copyin_simple
+
+end module m
+
+program main
+  use m, only: des_t, copyin_simple
+  implicit none
+
+  type(des_t) :: wdes1
+
+!$acc enter data copyin(wdes1)
+  call copyin_simple(wdes1)
+end program main
+
+! { dg-final { scan-tree-dump-not "\\.omp_data_arr\\..*\\.var = &var;" "omplower" } }
-- 
2.52.0

